packages/frontend-api/src/Model/Resolver/Products/ProductResolverMap.php

```diff
@@ -264,85 +303,87 @@ protected function mapProduct(): array
     {
         return [
```

Comment:
Had you considered to introduce some kind of serializer from `Product` to array which would have the same structure as elasticsearch index? Then you could check what data came into this only once and prepare array which is same for both ways. 🤔 
When I push this idea further it could result in view object (rather than magic array) like it is in `ReadModel` what do you think? It should not take too much of effort.

Reply:
Well, I see a problem with this approach. 
When I map a product to the array, I need to query all data and create the complete array. 

If I don't query categories (for example), the anonymous function in the map is not executed and this behavior cannot be achieved when I create the whole array beforehand.

Currently, I need to maintain both (ES and DB mapping) because, for example, PromotedProducts are still obtained from DB.

Reply:
refactored into multiple methods in 77e74b0cf

Reply:
refactored again into multiple classes in fd642fdf072893055899e162879a73a409ef1a1c